flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1234,6 +1235,14 @@ class ScalarExprLowering {
     if (resultTypes.size() == 1)
       resultType = resultTypes[0];
 
+    // TODO: find a nice way to drive how arguments should be lower to fir
+    // (address, value, or descriptor). For now, make a special case for
+    // inquiries, but driving how args are lowered with the intrinsic class
+    // will not be enough with all intrinsics.
```

Comment:
Will sufficiency be possible on an intrinsic by intrinsic basis or do the calling conventions of a distinct intrinsic also vary based on context?

Reply:
For most intrinsic, I think an intrinsic by intrinsic basis should be enough to define how each argument of the intrinsic should be lowered.

There may be special cases. For instance, with some inquiry intrinsic we may want to skip the argument computation (e.g in `SHAPE(MATMUL(A, B))`, there is an opportunity to avoid ever computing the MATMUL), so for those, what we do in lowering could depend on the context (unless of course, dead code elimination turns out good enough to remove the useless computation, and then it would not really matter to do something complex in lowering).

So far, I wanted to have the actual intrinsic implementation not depending on lib/evaluate to make it easy to create mlir ops for intrinsic and simply push back the implementation to a later pass. 

I think what I will probably do is to add a way in the intrinsic table to specify an optional callback that would take care of lowering the arguments from evaluate::Expr to ExtendedValues. That way, I can still enforce a split between the lowering and the "early fir implementation" parts.
I will need to give a try to more advanced inquiry/transformational to take a rational decision here.